### DISCUSSION TITLE
Execute vs. Invoke for resolving a Promise

### DISCUSSION THREAD

**Discussion by skicson** [sebastienros/jint#1121] [Q&A]
Hi, I'm new to Jint and have noticed that calling `Invoke` on JS code with a Promise does not resolve as expected but calling `Execute` does.  So given this code:
```         
            engine.SetValue("writeConsole", new Action<object>(Console.WriteLine));
 
            const string jscode = @"
                function _itest1() {
                    writeConsole('\nStart');
                    _itest2().then((res) => { writeConsole('Step 1, '+res);
                                              writeConsole('Finish');
                                            });
                }

                function _itest2() {
                    writeConsole('in _itest2');
                    return new Promise((resolve, reject) => {
                        resolve('Step 2');
                    });
                }
            ";

            engine.Execute(jscode);
            engine.Execute("_itest1()");
            engine.Invoke("_itest1");
         //   engine.Execute("writeConsole('foo');");
```
I get the following output

```
Start
in _itest2
Step 1, Step 2
Finish

Start
in _itest2

```
However, if I write to or flush the console, then the Invoke also finishes. 
```
            engine.Execute("_itest1()");
            engine.Invoke("_itest1");
            engine.Execute("writeConsole('foo');");
```

```
Start
in _itest2
Step 1, Step 2
Finish

Start
in _itest2
foo
Step 1, Step 2
Finish
```

Is this the expected behavior for `Invoke`?  I'm going to dig into the code to try and understand it better, but I'd appreciate any insights or suggestions.

**Comment by skicson** [comment]
It looks like the relevant difference between `engine.Invoke` and `engine.Execute` is that `engine.Execute` eventually calls `RunAvailableContinuations` but `engine.Invoke` does not.    The reason `Invoke` finishes in the second case above is because `RunAvailableContinuations` is called by the `engine.Execute("writeConsole('foo');");` but actually any `engine.Execute` call will work

**Comment by lahma** [comment]
I've never looked into `Invoke` that much, but based on my limited understanding it's more like for ad-hoc invoking of JS functions to get results whereas `Execute` and `Evaluate` are more suited for "full JS behavior".
